src/transformStateWithClones.js

```diff
@@ -7,7 +7,54 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+  const stateCopy = { ...state };
+  let changeState = stateCopy;
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        {
+          const addState = Object.assign({}, changeState, action.extraData);
+
+          result.push(addState);
+          changeState = addState;
+        }
+        break;
+
+      case 'removeProperties':
+        {
+          const arr = action.keysToRemove;
+          const removeState = { ...changeState };
+
+          for (const ch of arr) {
+            delete removeState[ch];
+          }
+
+          result.push(removeState);
+          changeState = removeState;
+        }
+        break;
+
+      case 'clear':
```

Comment:
For "clear", it's easier to just create a new empty object instead of a copy of the previous one.

Reply:
Thanks, I'll take it into account